homeassistant/components/sensorpush_cloud/api.py

```diff
@@ -0,0 +1,125 @@
+"""Library for interfacing with the SensorPush Cloud API."""
+
+from __future__ import annotations
+
+from asyncio import Lock
+from collections.abc import Awaitable, Callable, Coroutine, Mapping
+from datetime import datetime
+from functools import wraps
+from typing import Any, Concatenate
+
+from sensorpush_api import (
+    AccessTokenRequest,
+    ApiApi,
+    ApiClient,
+    AuthorizeRequest,
+    Configuration,
+    OpenApiException,
+    Samples,
+    SamplesRequest,
+    Sensor,
+    SensorsRequest,
+)
+
+from homeassistant.util import dt as dt_util
+
+from .const import ACCESS_TOKEN_EXPIRATION, LOGGER, REQUEST_RETRIES, REQUEST_TIMEOUT
+
+
+def api_call[**_P, _R](
+    func: Callable[Concatenate[SensorPushCloudApi, _P], Awaitable[_R]],
+) -> Callable[Concatenate[SensorPushCloudApi, _P], Coroutine[Any, Any, _R]]:
+    """Decorate API calls to handle SensorPush Cloud exceptions."""
+
+    @wraps(func)
+    async def _api_call(
+        self: SensorPushCloudApi, *args: _P.args, **kwargs: _P.kwargs
+    ) -> _R:
+        retries: int = 0
+        LOGGER.debug(f"API call to {func} with args={args}, kwargs={kwargs}")
+        while True:
+            try:
+                result = await func(self, *args, **kwargs)
+            except OpenApiException as e:
+                # Force reauthorization if an exception occurs to avoid
+                # authorization failures after temporary outages.
+                self.deadline = dt_util.now()
+
+                # The SensorPush Cloud API suffers from frequent exceptions;
+                # requests are retried before raising an error.
+                if retries < REQUEST_RETRIES:
+                    retries = retries + 1
+                    continue
+
+                LOGGER.debug(f"API call to {func} failed after {retries} retries")
+                raise SensorPushCloudError from e
+            else:
+                LOGGER.debug(f"API call to {func} succeeded after {retries} retries")
+                return result
+
+    return _api_call
+
+
+class SensorPushCloudError(Exception):
+    """An exception occurred when calling the SensorPush Cloud API."""
+
+
+class SensorPushCloudApi:
```

Comment:
I appreciate the feedback! I think there may be a bit of nuance missing. The API exposed by SensorPush is fairly large, though Home Assistant uses only a few endpoints. Because of this SensorPush opted to publish the API using Swagger rather than provide language-specific bindings. Moving this into a hand-written language binding would certainly remove the need for this integration code, however the amount of work it would take to create and maintain this over the long run doesn't make sense. Moving the integration code into the API would then mean that I would need to add this behavior to every endpoint to complete that interface, which doesn't make much sense either given the scope of the entire API.

With respect to email, this is not something that can change unless the user initiates an account change. There are no other keys that can be used to uniquely identify an account (the API uses email as the account identifier). I don't think there is much risk in using this as the unique ID for the service.

Given the above, if you still feel that this cannot be accepted into core, I understand. I was hoping that I could share this work with others - I have been contacted by someone already interested in making use of it. I can withdraw this PR if you wish and simply use this as a local custom integration. I'm more than happy to make other changes needed to merge the PR, but I'm afraid I have to draw the line somewhere and creating a bespoke language binding on top of a vendor-provided method of generating an API that has already been revised significantly to meet the guidelines of the feedback is too much.

Reply:
> Moving the integration code into the API would then mean that I would need to add this behavior to every endpoint to complete that interface, which doesn't make much sense either given the scope of the entire API.

I guess that is the downside of using a generated library. I maintain a few libraries that require low maintenance and are flexible enough to support such feature.

My experience with swagger generated libraries is that they have to be tweaked manually after updating more often than you'd like. Maybe there is a way you can patch this code in there. I think you can just add a module with this class to the library and be done with it?

I would love to see this integration in, but we also have to draw a line on service specific code.

Reply:
I completely agree. If I were using this for more than keeping an eye on my freezers I'd consider just making my own library. I spent some time this weekend looking through doing exactly that. Unfortunately OpenAPI doesn't have a central point where I can add the retry logic. I could add an additional module, but I would need to take what I've done and apply that to several more endpoints that I don't have a reliable way to test. Maybe an alternative would be to create a new library that just contains `api.py`, which in turn depends on `sensorpush-api`? This would keep the integration service code free without requiring a partial-stub in `sensorpush-api`. WDYT?

Reply:
I think that was proposed already, but then as a submodule rather than a separate dependency. I think both are viable options. If you go for the subdependency, both should be open source and licensed correctly

Reply:
Sounds good. I can have something ready to go this week. With respect to email, would you be okay with continuing to use that as a unique identifier?